
syntax = "proto3";

package lemming.dataplane.sai;

import "dataplane/standalone/proto/common.proto";

option go_package = "github.com/openconfig/lemming/proto/dataplane/sai";

enum LagAttr {
  LAG_ATTR_UNSPECIFIED = 0;
  LAG_ATTR_PORT_LIST = 1;
  LAG_ATTR_INGRESS_ACL = 2;
  LAG_ATTR_EGRESS_ACL = 3;
  LAG_ATTR_PORT_VLAN_ID = 4;
  LAG_ATTR_DEFAULT_VLAN_PRIORITY = 5;
  LAG_ATTR_DROP_UNTAGGED = 6;
  LAG_ATTR_DROP_TAGGED = 7;
  LAG_ATTR_TPID = 8;
  LAG_ATTR_SYSTEM_PORT_AGGREGATE_ID = 9;
  LAG_ATTR_LABEL = 10;
}

enum LagMemberAttr {
  LAG_MEMBER_ATTR_UNSPECIFIED = 0;
  LAG_MEMBER_ATTR_LAG_ID = 1;
  LAG_MEMBER_ATTR_PORT_ID = 2;
  LAG_MEMBER_ATTR_EGRESS_DISABLE = 3;
  LAG_MEMBER_ATTR_INGRESS_DISABLE = 4;
}

message CreateLagRequest {
  uint64 switch = 1;

  uint64 ingress_acl = 2;
  uint64 egress_acl = 3;
  uint32 port_vlan_id = 4;
  uint32 default_vlan_priority = 5;
  bool drop_untagged = 6;
  bool drop_tagged = 7;
  uint32 tpid = 8;
  uint32 system_port_aggregate_id = 9;
  bytes label = 10;
}

message CreateLagResponse {
  uint64 oid = 1;
}

message RemoveLagRequest {
  uint64 oid = 1;
}

message RemoveLagResponse {}

message SetLagAttributeRequest {
  uint64 oid = 1;
  oneof attr {
    uint64 ingress_acl = 2;
    uint64 egress_acl = 3;
    uint32 port_vlan_id = 4;
    uint32 default_vlan_priority = 5;
    bool drop_untagged = 6;
    bool drop_tagged = 7;
    uint32 tpid = 8;
    bytes label = 9;
  }
}

message SetLagAttributeResponse {}

message GetLagAttributeRequest {
  uint64 oid = 1;
  repeated LagAttr attr_type = 2;
}

message GetLagAttributeResponse {
  repeated LagAttribute attr = 1;
}

message CreateLagMemberRequest {
  uint64 switch = 1;

  uint64 lag_id = 2;
  uint64 port_id = 3;
  bool egress_disable = 4;
  bool ingress_disable = 5;
}

message CreateLagMemberResponse {
  uint64 oid = 1;
}

message RemoveLagMemberRequest {
  uint64 oid = 1;
}

message RemoveLagMemberResponse {}

message SetLagMemberAttributeRequest {
  uint64 oid = 1;
  oneof attr {
    bool egress_disable = 2;
    bool ingress_disable = 3;
  }
}

message SetLagMemberAttributeResponse {}

message GetLagMemberAttributeRequest {
  uint64 oid = 1;
  repeated LagMemberAttr attr_type = 2;
}

message GetLagMemberAttributeResponse {
  repeated LagMemberAttribute attr = 1;
}

service Lag {
  rpc CreateLag(CreateLagRequest) returns (CreateLagResponse) {}
  rpc RemoveLag(RemoveLagRequest) returns (RemoveLagResponse) {}
  rpc SetLagAttribute(SetLagAttributeRequest)
      returns (SetLagAttributeResponse) {}
  rpc GetLagAttribute(GetLagAttributeRequest)
      returns (GetLagAttributeResponse) {}
  rpc CreateLagMember(CreateLagMemberRequest)
      returns (CreateLagMemberResponse) {}
  rpc RemoveLagMember(RemoveLagMemberRequest)
      returns (RemoveLagMemberResponse) {}
  rpc SetLagMemberAttribute(SetLagMemberAttributeRequest)
      returns (SetLagMemberAttributeResponse) {}
  rpc GetLagMemberAttribute(GetLagMemberAttributeRequest)
      returns (GetLagMemberAttributeResponse) {}
}
